File: packages/javascript/lib/string/pad.ts
import type { PadOptions } from '../_types/global';

// Repeats the padding characters and cuts them off at `count`, so a multi-character `char`
// is used whole where it fits and truncated where it does not.
function buildPad(char: string, count: number): string {
	if (count <= 0) {
		return '';
	}

	const chars = Array.from(char);
	let result = '';

	for (let i = 0; i < count; i += 1) {
		result += chars[i % chars.length];
	}

	return result;
}

export function pad(str: string, length: number, options?: PadOptions): string {
	const text = str || '';
	const padChar = options?.char ?? ' ';
	// Counted in code points. A JavaScript string is indexed in UTF-16 units, so an emoji
	// would count as two here and as one in Python, and the result would differ per language.
	const textLength = Array.from(text).length;

	if (textLength >= length || !padChar) {
		return text;
	}

	const total = length - textLength;
	const position = options?.position ?? 'both';
	let startLength = Math.floor(total / 2);

	if (position === 'start') {
		startLength = total;
	} else if (position === 'end') {
		startLength = 0;
	}

	return buildPad(padChar, startLength) + text + buildPad(padChar, total - startLength);
}

File: packages/python/qsu/string/pad.py
def _buildPad(char: str, count: int) -> str:
	"""(Private) Repeat the padding characters and cut them off at `count`, so a
	multi-character `char` is used whole where it fits and truncated where it does not.
	"""
	if count <= 0:
		return ''

	return ''.join(char[i % len(char)] for i in range(count))


def pad(str, length, options=None, **kwargs):
	opts = {**(options or {}), **kwargs}
	text = str or ''
	padChar = opts.get('char', ' ')
	# Counted in code points, which is what Python does natively. JavaScript and Dart index
	# in UTF-16 units, so they walk the string by code point to answer the same way.
	textLength = len(text)

	if textLength >= length or not padChar:
		return text

	total = length - textLength
	position = opts.get('position', 'both')
	startLength = total // 2

	if position == 'start':
		startLength = total
	elif position == 'end':
		startLength = 0

	return _buildPad(padChar, startLength) + text + _buildPad(padChar, total - startLength)
